File: 1888.minimum-number-of-flips-to-make-the-binary-string-alternating.py
#
# @lc app=leetcode id=1888 lang=python3
#
# [1888] Minimum Number of Flips to Make the Binary String Alternating
#
#! a fucking hard and tricky problem
# @lc code=start
class Solution:
    def minFlips(self, s: str) -> int:
        ans, n = len(s), len(s)
        s += s
        #! Generate the two possible strings
        #? valid1: 01010101...
        #? valid2: 10101010...
        valid1 = ""
        valid2 = ""
        for i in range(2* n):
            if i % 2 == 0:
                valid1 += "0"
                valid2 += "1"
            else:
                valid1 += "1"
                valid2 += "0"
        #now we go through the for loop
        #How do I maintain the window?
        diff1, diff2 = 0, 0
        for i in range(2 * n):
            #Enter the window
            if s[i] != valid1[i]:
                diff1 += 1
            if s[i] != valid2[i]:
                diff2 += 1
            if i < n - 1:
                continue
            #This will be the first valid window of size n and we update the min Type2 flips
            ans = min(ans, min(diff1, diff2))
            #Now we Exit the window
            if s[i - n + 1] != valid1[i - n + 1]:
                diff1 -= 1
            elif s[i - n + 1] != valid2[i - n + 1]:
                diff2 -= 1
        return ans      
# @lc code=end

File: 1888.minimum-number-of-flips-to-make-the-binary-string-alternating.java
/*
 * @lc app=leetcode id=1888 lang=java
 *
 * [1888] Minimum Number of Flips to Make the Binary String Alternating
 */

// @lc code=start
class Solution {
    public int minFlips(String S) {
        // String s2 = s + s;
        // int One_pattern = 0;
        // int Zero_pattern = 0;
        // int ans = Integer.MAX_VALUE;
        // for(int i = 0; i < s2.length(); i++){
        //     if(i % 2 == 0){// Even position
        //         if (s2.charAt(i) == '0') One_pattern++;
        //         else Zero_pattern++;
        //     }
        //     else{
        //         if(s2.charAt(i) == '0') Zero_pattern++;
        //         else One_pattern++;
        //     }

        //     if(i < s.length() -1) continue;
        //     ans = Math.min(ans, Math.min(One_pattern, Zero_pattern));
        //     int j = i - s.length() + 1;
        //     if(j % 2 == 0){
        //         if (s2.charAt(i - s.length() + 1) == '0') One_pattern--;
        //         else Zero_pattern--;
        //     }
        //     else{
        //         if(s2.charAt(i - s.length() + 1) == '0') Zero_pattern--;
        //         else One_pattern--;
        //     }
        // }
        // return ans; 

        //! solid thinking and systematic knowledgebase are the key to leetcode
        
        char[] s = S.toCharArray();
        int n = s.length;
        int ans = n;
        int cnt = 0;
        for(int i = 0; i < 2*n - 1; i++){
            if(s[i % n] % 2 != i % 2) cnt++;
            int left = i - n + 1;
            if(left < 0) continue;
            ans = Math.min(ans, Math.min(cnt, n - cnt));
            if(s[left] % 2 != left % 2) cnt--;
        }
        return ans;
    }
}
// @lc code=end

